PLMS.BLL/ServicesImplementation/TaskService.cs

```diff
@@ -1,206 +1,213 @@
-﻿using PLMS.BLL.DTO;
+﻿using AutoMapper;
+using Microsoft.EntityFrameworkCore;
+using PLMS.BLL.DTO.TasksDto;
+using PLMS.BLL.Filters;
 using PLMS.BLL.ServicesInterfaces;
+using PLMS.Common.Exceptions;
 using PLMS.DAL.Entities;
-using PLMS.BLL.Filters;
 using PLMS.DAL.Interfaces;
+using System.Linq.Expressions;
 using Task = PLMS.DAL.Entities.Task;
 
 namespace PLMS.BLL.ServicesImplementation
 {
     public class TaskService : ITaskService
     {
-
         private readonly IUnitOfWork _unitOfWork;
 
+        private readonly IMapper _mapper;
         private readonly IRepository<Task> _taskRepository;
+        private readonly IRepository<Goal> _goalRepository;
 
-        public TaskService(IUnitOfWork unitOfWork)
+        public TaskService(IUnitOfWork unitOfWork, IMapper mapper)
         {
             _unitOfWork = unitOfWork;
+            _mapper = mapper;
             _taskRepository = _unitOfWork.GetRepository<Task>();
+            _goalRepository = _unitOfWork.GetRepository<Goal>();
         }
-        public async System.Threading.Tasks.Task AddTaskAsync(TaskDto taskDto)//ToDo: should be a separate DTO class for adding Task
+
+        public async Task<int> AddTaskAsync(AddTaskDto taskDto)
         {
-            Task task = new()
+            var goal = await _goalRepository.GetByIdAsync(taskDto.GoalId) ?? throw new NotFoundException("Goal was not found");
+            if (goal != null && goal.UserId != taskDto.UserId)
             {
-                Title = taskDto.Title,
-                Description = taskDto.Description,
-                //GoalId = taskDto.GoalId,//you can use this for linking existing Goal
-                //Goal = taskDto.GoalDto, //you can use this for creating a new Goal, don't forget to map it from DTO to EF entity
-                //Category = taskDto.CategoryDto,//same
-                //Priority = taskDto.PriorityDto//same
-                CreatedAt = taskDto.CreatedAt,//TODO: User should not be able to set this field, it should not be present in the DTO, it should be autofilled here instead
-                DueDate = taskDto.DueDate,
-                TaskComments = taskDto.Comments.Select(x => new TaskComment { Comment = x.Comment, CreatedAt = x.CreatedAt}).ToArray()
-            };
+                throw new UnauthorizedAccessException("Can't add this task!");
```

Comment:
In all other methods you return `NotFoundException("Task was not found");` in case when permissions denied.
Therefore, make this method consistent and return NotFoundException("Goal was not found"); instead of UnauthorizedAccessException.